drv/lpc55-swd/src/main.rs

```diff
@@ -0,0 +1,641 @@
+        }
+    }
+}
+
+impl ServerImpl {
+    fn io_out(&mut self) {
+        while !self.spi.mstidle() {
+            cortex_m::asm::nop();
+        }
+
+        let gpio_driver = GPIO.get_task_id();
+        switch_io_out(gpio_driver);
+    }
+
+    fn io_in(&mut self) {
+        while !self.spi.mstidle() {
+            cortex_m::asm::nop();
+        }
+
+        let gpio_driver = GPIO.get_task_id();
+        switch_io_in(gpio_driver);
+    }
+
+    fn read_ack(&mut self) -> Result<(), Ack> {
+        // This read includes the turnaround bit which we
+        // don't care about.
+        let b = self.read_nibble();
+
+        // We configured the SPI controller to give us back 4 bits,
+        // if we got more than that something has gone very wrong
+        if b & 0xF0 != 0 {
+            ringbuf_entry!(Trace::AckErr(Ack::Protocol));
+            return Err(Ack::Protocol);
+        }
+
+        // Section 5.3 of ADIv5 describes the bit patterns
+        match b & 0x7 {
+            0b001 => {
+                ringbuf_entry!(Trace::AckErr(Ack::Fault));
+                Err(Ack::Fault)
+            }
+            0b010 => {
+                ringbuf_entry!(Trace::AckErr(Ack::Wait));
+                Err(Ack::Wait)
+            }
+            0b100 => Ok(()),
+            _ => {
+                ringbuf_entry!(Trace::AckErr(Ack::Protocol));
+                Err(Ack::Protocol)
+            }
+        }
+    }
+
+    fn tx_byte(&mut self, byte: u8) {
+        while !self.spi.can_tx() {}
+        self.spi.send_u8_no_rx(byte);
+    }
+
+    // SW-DP is intended to be used as a bit based protocol.
+    // The smallest unit the SPI controller can do is 4 bits
+    fn read_nibble(&mut self) -> u8 {
+        while !self.spi.can_tx() {}
+        self.spi.send_raw_data(0x0, true, false, 4);
+        while !self.spi.has_byte() {}
+        self.spi.read_u8()
+    }
+
+    fn read_trailing(&mut self) -> u8 {
+        while !self.spi.can_tx() {}
+        self.spi.send_raw_data(0x0, true, false, 5);
+        while !self.spi.has_byte() {}
+        self.spi.read_u8()
+    }
+
+    fn swd_transfer_cmd(
+        &mut self,
+        port: Port,
+        reg: RawSwdReg,
+    ) -> Result<(), Ack> {
+        self.io_out();
+
+        // has our start and stop bits set
+        let mut byte: u8 = START_VAL | PARK_VAL;
+
+        let (rd, abits) = get_addr_and_rw(reg);
+
+        let port_bit: u8 = match port {
+            Port::DP => (0 << APDP_BIT),
+            Port::AP => (1 << APDP_BIT),
+        };
+
+        byte |= abits | rd | port_bit;
+
+        let p = calc_parity(byte);
+
+        byte |= p;
+
+        self.tx_byte(byte);
+
+        self.io_in();
+
+        self.read_ack()
+    }
+
+    fn reset(&mut self) {
+        // Spec says hold high for 51 clock cycles, more is okay
+        // this gives us 56
+        for _ in 0..7 {
+            self.tx_byte(0xff);
+        }
+    }
+
+    fn idle_cycles(&mut self, cnt: usize) {
+        for _ in 0..(cnt >> 3) {
```

Comment:
Took me a second to realize (cnt >> 3) is bits->bytes conversion.  I would have recognized (cnt/8) faster.  If nothing else, a comment about bits->bytes may be warranted.

Reply:
Yes, will update with a comment and also round up just to be cautious.